src/reducers/entities.spec.js

```diff
@@ -0,0 +1,101 @@
+import entityReducer from './entities'
+import { updateEntity, receiveEntities } from 'actions/entities'
+
+describe('entityReducer', () => {
```

Comment:
Ideally, tests should make easier understand what the code does, where the code is used, and what are corner cases of those use cases.

For doing that, we need to make them more readable. Somethin like should work:
```
// entities.spec.js  <- Good name. it specifies what we are testing
describe('Reducer Entities: Adding users')  <-- Should describe a feature of the domain we are testing
    test("Adding users when the store is empty")   <-- Should specify all different cases
        ...
    test("Adding users different users")
        ...
    test("Modifying existing users")  <-- This test is an example of the typical test added later in time
                                                                  because we have had a bug, like mutations
        //GIVEN
             const storeWithUsers  //names are just an example
             const dataFromEndpointWithSameUsers  //names are just an example
        //WHEN
             //adding to the store              
        //THEN
            //We do not have mutations 
```
If you follow this pattern, you will have better tests, more organised and new developers will understand everything much better and faster.

Reply:
Thanks for the feedback, can you maybe give me a concrete example of how this should look for this reducer?

I rewrote it like this, wdyt?
```jsx
describe('Reducer Entities: Adding and updating Entities', () => {
  describe('Action: updateEntity', () => {
    // GIVEN
    const stateWithTestEntity = {
      testEntity: {
        'test-2': {
          id: 'test-2',
          deeplyNestedData: {
            deeplyNestedKey: 'moreNestedData',
          },
        },
      },
    }

    // WHEN
    const updateAction = updateEntity({ entityType: 'testEntity', data: { id: 'test-1', testObject: { deeplyNested: 'aNestedVaue' } } })
    const stateWithUpdatedEntity = entityReducer(stateWithTestEntity, updateAction)

    // THEN
    test('Updating an Entity', () => {
      // Ensure Action Payload matches Entity Payload
      expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).toMatchObject(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
      expect(updateAction.payload.data.testObject).toMatchObject(stateWithUpdatedEntity.testEntity['test-1'].testObject)
      
      // Ensure no mutations! (=== Compare)
      expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).not.toBe(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
      expect(updateAction.payload.data.testObject).not.toBe(stateWithUpdatedEntity.testEntity['test-1'].testObject)

      // Ensure new Entity was added
      const stateWithNewEntity = {
        'test-2': {
          id: 'test-2',
          deeplyNestedData: {
            deeplyNestedKey: 'moreNestedData',
          },
        },
        'test-1': {
          id: 'test-1',
          testObject: { deeplyNested: 'aNestedVaue' },
        },
      }
      expect(stateWithUpdatedEntity.testEntity).toMatchObject(stateWithNewEntity)
    })
  })
```

Reply:
Ok, let me ask first a couple of questions:
+  Why do you think it is a good idea to have chained ```describe```? 
+  Which entities are you testing? Users? Trades? Markets?

Reply:
- I'm testing a reducer and inside this reducer, the two specified actions it can receive
- I'm testing the entity reducer, so I'm using an entity called "TestEntity"

Reply:
@apanizo 
Okay, so does this make more sense then?
```javascript
describe('Reducer Entities: Adding and updating Entities', () => {
  test('Updating an Entity', () => {
    // GIVEN
    const stateWithTestEntity = {
      testEntity: {
        'test-2': {
          id: 'test-2',
          deeplyNestedData: {
            deeplyNestedKey: 'moreNestedData',
          },
        },
      },
    }

    // WHEN
    const updateAction = updateEntity({ entityType: 'testEntity', data: { id: 'test-1', testObject: { deeplyNested: 'aNestedVaue' } } })
    const stateWithUpdatedEntity = entityReducer(stateWithTestEntity, updateAction)

    // THEN
    // Ensure Action Payload matches Entity Payload
    expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).toMatchObject(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
    expect(updateAction.payload.data.testObject).toMatchObject(stateWithUpdatedEntity.testEntity['test-1'].testObject)
    
    // Ensure no mutations! (=== Compare)
    expect(stateWithTestEntity.testEntity['test-2'].deeplyNestedData).not.toBe(stateWithUpdatedEntity.testEntity['test-2'].deeplyNestedData)
    expect(updateAction.payload.data.testObject).not.toBe(stateWithUpdatedEntity.testEntity['test-1'].testObject)

    // Ensure new Entity was added
    const stateWithNewEntity = {
      'test-2': {
        id: 'test-2',
        deeplyNestedData: {
          deeplyNestedKey: 'moreNestedData',
        },
      },
      'test-1': {
        id: 'test-1',
        testObject: { deeplyNested: 'aNestedVaue' },
      },
    }
    expect(stateWithUpdatedEntity.testEntity).toMatchObject(stateWithNewEntity)
  })
```

Reply:
Testing should be done based on features. We can say a feature is something that the user can do in the application.

Orienting tests in that way will help to understand what the code should do and what we expect from that code.  So Testing a reducer is not something a user can do.

So the question is... what the user can do for triggering the usage of that reducer? Maybe load markets and therefore "Add Markets" to the store? Maybe delete a market? 

With that information, I will compose the "describe" functions. For example:
```
describe('Add Markets to store')
describe('Remove markets to store')
```
Once I have the features _described_ the next step is to write the different tests cases. For example:
Which different options can we have when we add markets?
  * Adding markets when the store is empty ```test('Adding markets when store is empty`)```
  * Adding different markets that the ones who are inside the store
  * Finally modify existing markets

The last point is (or the last test case) is the typical test you do not add at the beginning, is the typical test you add when you have had a bug (like mutations).

The last step is how to organize the test. For reading purposes I suggest to use:
```
GIVEN
WHEN
THEN
```
approach.

For example:
```
describe('Add Markets') { 
    test('Add Markets when store is empty') {
        //GIVEN
        const emptyStore = ...;
        const payload = request.payload;

        //WHEN... adding markets to store
        const store = {...emptyStore, payload.markets }

       //THEN... we have the markets in the store
        const markets = getMarketSelector()
       testEqual(market[0], payload.market[0]);
       testEqual(market[1], payload.market[1]);
       ....
    }
    ....
}
describe(' ...') {
    ....
}
.....
```
So I do not think having chained is good in terms of understanding correctly what the app does.